[MXNET-1121] Example to demonstrate the inference workflow using RNN (#13680)
apache/incubator-mxnet: cpp-package/example/inference/README.md

```
./unit_test_inception_inference.sh
```

### [simple_rnn.cpp](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/simple_rnn.cpp>)

> This example demonstrates sequence prediction workflow with pre-trained RNN model using MXNet C++ API. The purpose of this example is to demonstrate how pre-trained RNN model can be loaded and used to generate an output sequence using C++ API.

Author: Fixed.

> The example performs following tasks.

Author: Fixed.

> - Load the pre-trained RNN model,

Author: Fixed.

- Load the dictionary file that contains word to index mapping.
- Convert the input string to vector of indices and padded to match the input data length.
- Run the forward pass and predict the output string.

The example uses a pre-trained RNN model that is trained with the dataset containing speeches given by Obama. The model files can be found here.
- [obama-speaks-symbol.json](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-symbol.json>)
- [obama-speaks-0100.params](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-0100.params>)

> - [obama.dictionary.txt](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama.dictionary.txt>) Each line of the dictionary file contains a word and an unique index for that word separated by a space. This dictionary file is created when the RNN model was trained with a particular dataset.

Author: Fixed.

> The example downloads the above files while running. The example is intentionally kept specific to a certain model because:

Author: Fixed.

- Unavailability of pre-trained RNN models that can be successfully imported using C++.
- C++ API currently does not support bucketing. Hence the shape of input data needs to be fixed and known while loading the model.

> **TODO**: The example can later be enhanced once the above limitations are addressed.

Author: Done.

```
./unit_test_inception_inference.sh
```

### [simple_rnn.cpp](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/simple_rnn.cpp>)
This example demonstrates sequence prediction workflow with pre-trained RNN model using MXNet C++ API. The purpose of this example is to demonstrate how a pre-trained RNN model can be loaded and used to generate an output sequence using C++ API.
The example performs following tasks

> - Load the pre-trained RNN model.

Reviewer: ```suggestion
- Loads the pre-trained RNN model.
```

> - Load the dictionary file that contains word to index mapping.

Reviewer: ```suggestion
- Loads the dictionary file containing the word to index mapping.
```

> - Convert the input string to vector of indices and padded to match the input data length.

Reviewer: ```suggestion
- Converts the input string to vector of indices that's padded to match the input data length.
```

> - Run the forward pass and predict the output string.

Reviewer: ```suggestion
- Runs the forward pass and predicts the output string.
```

> The example uses a pre-trained RNN model that is trained with the dataset containing speeches given by Obama.

Reviewer: ```suggestion
The example uses a pre-trained RNN model trained with a dataset containing speeches given by Obama.
```

The model consists of :

> - Embedding Layer with the size of embedding to be 650

Author: I can get the pdf version of model (generated using visualiation API). Need some suggestions from @aaronmarkham  to embed them in README.

> - Embedding Layer with the size of embedding to be 650

Reviewer: ```suggestion
- Embedding Layer with the size of embedding set at 650
```

- 3 LSTM Layers with hidden dimension size of 650 and sequence length of 35
- FullyConnected Layer
- SoftmaxOutput
The model was trained for 100 epochs.

The model files can be found here.
- [obama-speaks-symbol.json](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-symbol.json>)

> - [obama-speaks-0100.params](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama-speaks-0100.params>)

Author: The current bucket mxnet-cpp is not a public by default but the contents are made publicly readable. This is similar to mxnet-scala bucket used for scala examples.

- [obama.dictionary.txt](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/obama.dictionary.txt>) Each line of the dictionary file contains a word and a unique index for that word, separated by a space, with a total of 14293 words generated from the training dataset.
The example downloads the above files while running.

The example's command line parameters are as shown below:

```
./simple_rnn --help
Usage:
simple_rnn
[--input] Input string sequence.
[--gpu]  Specify this option if workflow needs to be run in gpu context.

./simple_rnn

or

./simple_rnn --input "Good morning. I appreciate the opportunity to speak here"
```

> The example will output the seqence of 35 words as follows:

Author: Done.

```

> [waters elected Amendment Amendment Amendment Amendment retirement maximize maximize maximize acr citi sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio ]

Author: I tried to get better output by changing model hyperparameters but couldn't get it. It would require a good amount of input data processing as well. All these efforts would require dedicated time and out of scope for this example. The example aims towards loading the model and running forward pass. Improving on the model would be a separate task.

> [waters elected Amendment Amendment Amendment Amendment retirement maximize maximize maximize acr citi sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio sophisticatio ]

Author: I am working on implementing the RNN model using C++ API. I can work on improving the accuracy of that model and use it in this example later.

```

> Alternatively, user can run [unit_test_simple_rnn.sh](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/unit_test_simple_rnn.sh>) script.

Reviewer: ```suggestion
Alternatively, you can run the [unit_test_simple_rnn.sh](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/unit_test_simple_rnn.sh>) script.
```

The model consists of :
- Embedding Layer
- 2 LSTM Layers with hidden dimension size of 200
- Average pooling layer
- Sigmoid output layer
The model was trained for 10 epochs to achieve 85% test accuracy.
The visual representation of the model is [here](<http://gluon-nlp.mxnet.io/examples/sentiment_analysis/sentiment_analysis.html#Sentiment-analysis-model-with-pre-trained-language-model-encoder>).

The model files can be found here.
- [sentiment_analysis-symbol.json](< https://s3.amazonaws.com/mxnet-cpp/RNN_model/sentiment_analysis-symbol.json>)
- [sentiment_analysis-0010.params](< https://s3.amazonaws.com/mxnet-cpp/RNN_model/sentiment_analysis-0010.params>)
- [sentiment_token_to_idx.txt](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/sentiment_token_to_idx.txt>) Each line of the dictionary file contains a word and a unique index for that word, separated by a space, with a total of 32787 words generated from the training dataset.
The example downloads the above files while running.

The example's command line parameters are as shown below:

```
./sentiment_analysis_rnn --help
Usage:
sentiment_analysis_rnn
--input Input movie review line.e.g. "This movie is the best"

> [--max_num_words]  The number of words in the sentence to be considered for sentiment analysis. Default is 5

Reviewer 2: why 5? I think a usual sentence will be longer than this.

Reviewer 2: - I suggest adding more examples of different lengths so it will be clear for users.
- Can we make it clear that if input is less than the length of max_num_words it gets padded with 0s and if it is longer it gets trimmed.

Author: Thanks @sandeep-krishnamurthy  for review.
I have increased the length to 10. This can also be specified as a command line parameters.
I have included examples of various lengths in README file.
The "Help" message is updated.

- Average pooling layer
- Sigmoid output layer
The model was trained for 10 epochs to achieve 85% test accuracy.
The visual representation of the model is [here](<http://gluon-nlp.mxnet.io/examples/sentiment_analysis/sentiment_analysis.html#Sentiment-analysis-model-with-pre-trained-language-model-encoder>).

The model files can be found here.
- [sentiment_analysis-symbol.json](< https://s3.amazonaws.com/mxnet-cpp/RNN_model/sentiment_analysis-symbol.json>)
- [sentiment_analysis-0010.params](< https://s3.amazonaws.com/mxnet-cpp/RNN_model/sentiment_analysis-0010.params>)
- [sentiment_token_to_idx.txt](<https://s3.amazonaws.com/mxnet-cpp/RNN_model/sentiment_token_to_idx.txt>) Each line of the dictionary file contains a word and a unique index for that word, separated by a space, with a total of 32787 words generated from the training dataset.
The example downloads the above files while running.

The example's command line parameters are as shown below:

```
./sentiment_analysis_rnn --help
Usage:
sentiment_analysis_rnn
--input Input movie review line.e.g. "This movie is the best." The input is trimmed or padded with 0s to match the max_num_words.

> [--max_num_words]  The number of words in the sentence to be considered for sentiment analysis. Default is 10

Reviewer 3: Is there any upper bound for ```max_num_words```?

- FullyConnected Layer
- SoftmaxOutput
The model was trained for 100 epochs.

> The visual representation of the model is [here](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference/obama-speaks.pdf>).

Author: Thanks! I have added it now.
